### PR TITLE
Needle's Star Road Fixes (12/14/2024)

### PR DESCRIPTION
**Pull request by 64BitMouse**
- You now get sent to Zero Life after dying with a life count below Zero instead of at Zero

- Tuxie (Race Penguin) No longer clips through geometry when finishing the race in Colossal Candy Clutter

- Set Number of lives from 4 to 3 to be more accurate

- Implemented a fully functional Hard Mode as the original hack also had one (Hold L as you host Star Road OR Activate it through the mod menu. Uses the backup save slot.)

- Koopa the Quick in both Koopa Canyon and Mushroom Mountain Town no longer softlocks himself

- Added a signpost in Castle Grounds hinting at Hard Mode

- Fixed a strange warp node in Zero Life to warp you to the start warp

- Moved a Goomba farther from the Warp Pipe in Hidden Palace Finale to prevent him from hitting you immediately (Also makes Hard Mode Possible)

- Fixed SMSRSpaceBox's behavior as it was initially missing the SET_HOME() Function (Was basically snapping to Level Origin so we had to use Pushable Metal Boxes to make "Climb Through the Machinery" (Starlight Runway - Star 2) Casually Possible

### CODE REVIEW

**Comment by Flower35** [comment]
Please change the target branch to **`coop-deluxe:dev`** when openin a Pull Request.

File: mods/star-road/main.lua [hunk 1]
```diff
@@ -104,13 +115,39 @@ smlua_audio_utils_replace_sequence(0x32, 0x25, 80, "32_Seq_smsrdeluxe_custom")
...
-    if m.numLives <= 0 then
-        warp_to_level(LEVEL_ZERO_LIFE, 1, 0)
-        m.numLives = 4
+    if m.numLives < 0 and transitionType == WARP_TRANSITION_FADE_INTO_BOWSER then
+        zerolife = true
+        m.numLives = 3
+    end
+end
+
+function  void_death()
```
Comment (PeachyPeachSM64) [inline mods/star-road/main.lua]: 2 spaces

File: mods/star-road/main.lua [hunk 2]
```diff
@@ -104,13 +115,39 @@ smlua_audio_utils_replace_sequence(0x32, 0x25, 80, "32_Seq_smsrdeluxe_custom")
...
+    local m = gMarioStates[0]
+    if m.numLives < 0  then
```
Comment (PeachyPeachSM64) [inline mods/star-road/main.lua]: 2 spaces

File: mods/star-road/main.lua [hunk 3]
```diff
@@ -104,13 +115,39 @@ smlua_audio_utils_replace_sequence(0x32, 0x25, 80, "32_Seq_smsrdeluxe_custom")
...
+    if m.playerIndex ~= 0 then return end
+    if startingLives then
+        m.numLives = 3
+        startingLives = false
+    end
+    if zerolife then
+        warp_to_level(LEVEL_ZERO_LIFE, 1, 0) 
+        zerolife = false
+        startingLives = true
+
```
Comment (PeachyPeachSM64) [inline mods/star-road/main.lua]: unneeded newline

File: mods/star-road/main.lua [hunk 4]
```diff
@@ -1,41 +1,52 @@
 -- name: Star Road
--- description: Star Road is a romhack created by SKELUX\n\nwww.youtube.com/c/Skelux\n\nThere are 23 custom levels, with a total of 130 stars.\n\nPorted to PC by AloXado320, PeachyPeach, and PrinceFrizzy\n\nPorted to coop by djoslin0, PrinceFrizzy, and Isaac.
+-- description: Romhack created by \\#FFD700\\SKELUX\\#DCDCDC\\\n\nwww.youtube.com/c/Skelux\n\nThere are 23 custom levels, with a total of 130 stars.\n\nPorted to PC by AloXado320,\n\\#C96BBA\\PeachyPeach\\#DCDCDC\\, and \\#8C4AC7\\PrinceFrizzy\\#DCDCDC\\\n\nPorted to coop by \\#093614\\djoslin0\\#DCDCDC\\,\n\\#8C4AC7\\PrinceFrizzy\\#DCDCDC\\, and \\#141699\\Isaac\\#DCDCDC\\.\n\nAdditional Fixes and "\\#FF0000\\Hard Mode\\#DCDCDC\\" addition by  \\#2D851C\\Mr.Needlemouse\\#DCDCDC\\.
```
Comment (PeachyPeachSM64) [inline mods/star-road/main.lua]: please replace `#C96BBA` by `#FFAAFF`

File: mods/star-road/hardmode.lua [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    if not gGlobalSyncTable.hardMode then return false end
+end
+
+function hard_mode_menu(_, value)
+    gGlobalSyncTable.hardMode = value
+    save_file_set_using_backup_slot(gGlobalSyncTable.hardMode)
+    save_file_reload(1)
+    djui_popup_create(value and "\\#ff0000\\HARD MODE\\#dcdcdc\\ Activated" or "You are currently in NORMAL MODE.", 1)
+    play_sound(SOUND_MENU_MARIO_CASTLE_WARP + 1 << 0, { x = 0, y = 0, z = 0 })
+    warp_to_warpnode(16, 1, 0, 128)
```
Comment (PeachyPeachSM64) [inline mods/star-road/hardmode.lua]: you should replace `16` by `LEVEL_CASTLE_GROUNDS`

File: mods/star-road/hardmode.lua [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    end
+    if not gGlobalSyncTable.hardMode then return false end
+end
+
+function hard_mode_menu(_, value)
+    gGlobalSyncTable.hardMode = value
+    save_file_set_using_backup_slot(gGlobalSyncTable.hardMode)
+    save_file_reload(1)
+    djui_popup_create(value and "\\#ff0000\\HARD MODE\\#dcdcdc\\ Activated" or "You are currently in NORMAL MODE.", 1)
+    play_sound(SOUND_MENU_MARIO_CASTLE_WARP + 1 << 0, { x = 0, y = 0, z = 0 })
```
Comment (PeachyPeachSM64) [inline mods/star-road/hardmode.lua]: what is that `+ 1 << 0` supposed to do?

File: mods/star-road/hardmode.lua [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    if not gGlobalSyncTable.hardMode then return end
+    if not seenPopup and not network_is_server() then
+        save_file_set_using_backup_slot(gGlobalSyncTable.hardMode)
+        save_file_reload(1)
+        djui_popup_create("You have joined a \\#ff0000\\HARD MODE\\#dcdcdc\\ lobby.", 3)
+        play_sound(SOUND_MENU_MARIO_CASTLE_WARP + 1 << 0, { x = 0, y = 0, z = 0 })
+        seenPopup = true
+    end
+    if m.hurtCounter > 0 or burnActions[m.action] then
+        m.hurtCounter = 64 -- Damage Check. Hard Mode is a OHKO mode.
```
Comment (PeachyPeachSM64) [inline mods/star-road/hardmode.lua:42]: why not instantly kill Mario by setting his health to `0xFF` whenever he takes damage?
Also I do recall that og star road hard mode gives Mario infinite breath underwater; is it intended that this feature doesn't appear here?

**Review by PeachyPeachSM64 — COMMENTED** [review]
Can you also provide the changes made to the source of the behavior and level files?

As a side note, I think we should add the behavior and level source files for both sm74 and star road to this repository. It would be easier to maintain and review these two rom-hacks.

File: mods/star-road/hardmode.lua (hunk 7 above)
Comment (64BitMouse) [reply mods/star-road/hardmode.lua:42]: I didnt think the water thing was necessary but I can add it I suppose

Comment (64BitMouse) [reply mods/star-road/hardmode.lua:42]: I also set the hurt counter instead of just setting health to 0xFF as more of a visual thing.

**Pull request closed without merging**